File: app/src/main/java/org/scp/gymlog/ui/common/components/NumberModifierView.java
package org.scp.gymlog.ui.common.components;

import static org.scp.gymlog.util.FormatUtils.toBigDecimal;
import static org.scp.gymlog.util.FormatUtils.toDp;
import static org.scp.gymlog.util.FormatUtils.toDpFloat;

import android.annotation.SuppressLint;
import android.content.Context;
import android.util.AttributeSet;
import android.util.DisplayMetrics;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.MotionEvent;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;

import androidx.annotation.DrawableRes;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;

import org.scp.gymlog.R;
import org.scp.gymlog.util.FormatUtils;

import java.math.BigDecimal;

public class NumberModifierView extends LinearLayout {
    private int targetId = -1;
    private BigDecimal step = BigDecimal.ONE;
    private boolean allowNegatives = false;

    private boolean autoExecuted = false;
    private int autoCount = 0;
    private Runnable autoAdd;
    private Runnable autoSub;

    public NumberModifierView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        if (attrs != null) {
            targetId = attrs.getAttributeResourceValue(
                    "http://schemas.android.com/apk/res-auto",
                    "target",
                    -1);

            allowNegatives = attrs.getAttributeBooleanValue(
                    "http://schemas.android.com/apk/res-auto",
                    "negatives",
                    false);

            step = BigDecimal.valueOf(attrs.getAttributeFloatValue(
                    "http://schemas.android.com/apk/res-auto",
                    "step",
                    1f));
        }

        autoAdd = () -> {
            autoExecuted = true;
            onAction(true);
            postDelayed(autoAdd, ++autoCount > 30? 25 : 55);
        };

        autoSub = () -> {
            autoExecuted = true;
            onAction(false);
            postDelayed(autoSub, ++autoCount > 30? 25 : 55);
        };

        addView();
    }

    public NumberModifierView setStep(BigDecimal step) {
        this.step = step;
        return this;
    }

    public NumberModifierView allowNegatives() {
        this.allowNegatives = true;
        return this;
    }

    private void addView() {
        CardView subCv = createCardView();
        LayoutParams params = (LayoutParams) subCv.getLayoutParams();
        params.setMarginEnd(toDp(getResources().getDisplayMetrics(), 3));
        subCv.addView(createImageView(R.drawable.ic_substract_24dp));
        modifyEditText(subCv, false);
        addView(subCv);

        CardView addCv = createCardView();
        addCv.addView(createImageView(R.drawable.ic_add_24dp));
        modifyEditText(addCv, true);
        addView(addCv);
    }

    private CardView createCardView() {
        CardView cardView = new CardView(getContext());
        DisplayMetrics displayMetrics = getResources().getDisplayMetrics();

        int size = toDp(displayMetrics, 50);
        LayoutParams layoutParams = new LayoutParams(size, size);
        cardView.setLayoutParams(layoutParams);

        int color = getContext().getResources().getColor(
                R.color.dark,
                getContext().getTheme());

        cardView.setCardBackgroundColor(color);
        cardView.setRadius(toDpFloat(displayMetrics, 3));
        cardView.setCardElevation(0);

        return cardView;
    }

    private ImageView createImageView(@DrawableRes int drawable) {
        ImageView imageView = new ImageView(getContext());

        int size = toDp(getResources().getDisplayMetrics(), 22);
        CardView.LayoutParams layoutParams = new CardView.LayoutParams(size, size);
        layoutParams.gravity = Gravity.CENTER;
        imageView.setLayoutParams(layoutParams);

        imageView.setImageResource(drawable);

        int color = getContext().getResources().getColor(
                R.color.white,
                getContext().getTheme());
        imageView.setColorFilter(color);

        return imageView;
    }

    @SuppressLint("ClickableViewAccessibility")
    private void modifyEditText(CardView cardView, boolean addition) {
        cardView.setOnTouchListener((view, event) -> {
            switch (event.getAction()) {
                case MotionEvent.ACTION_DOWN:
                    if (addition) {
                        removeCallbacks(autoAdd);
                        autoCount = 0;
                        postDelayed(autoAdd, 300);
                    } else {
                        removeCallbacks(autoSub);
                        autoCount = 0;
                        postDelayed(autoSub, 300);
                    }
                    return true;

                case MotionEvent.ACTION_UP:
                    removeCallbacks(addition? autoAdd : autoSub);
                    if (autoExecuted) {
                        autoExecuted = false;
                    } else {
                        onAction(addition);
                    }
                    return true;
            }
            return false;
        });
    }


    private void onAction(boolean addition) {
        EditText editText = getRootView().findViewById(targetId);
        if (editText != null) {
            BigDecimal value = toBigDecimal(editText.getText().toString());
            BigDecimal step = addition? this.step : this.step.negate();

            value = value.add(step);
            if (!allowNegatives && value.compareTo(BigDecimal.ZERO) <= 0)
                value = BigDecimal.ZERO;

            editText.setText(FormatUtils.toString(value));
            editText.setSelection(editText.getText().length());
        }
    }
}

File: app/src/main/java/org/scp/gymlog/ui/common/components/NumberModifierView.kt
package org.scp.gymlog.ui.common.components

import android.annotation.SuppressLint
import android.content.Context
import android.util.AttributeSet
import android.view.Gravity
import android.view.MotionEvent
import android.widget.EditText
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import androidx.annotation.DrawableRes
import androidx.cardview.widget.CardView
import org.scp.gymlog.R
import org.scp.gymlog.util.FormatUtils
import org.scp.gymlog.util.FormatUtils.bigDecimal
import java.math.BigDecimal

class NumberModifierView(context: Context, attrs: AttributeSet) : LinearLayout(context, attrs) {

    companion object {
        private const val NAMESPACE = "http://schemas.android.com/apk/res-auto"
    }

    private val targetId: Int
    private var step: BigDecimal
    private var allowNegatives: Boolean
    private var autoExecuted = false
    private var autoCount = 0
    private val autoAdd: Runnable
    private val autoSub: Runnable

    init {
        targetId = attrs.getAttributeResourceValue(
            NAMESPACE, "target", -1)

        allowNegatives = attrs.getAttributeBooleanValue(
            NAMESPACE, "negatives", false)

        step = attrs.getAttributeFloatValue(
            NAMESPACE, "step", 1f)
            .toBigDecimal()

        autoAdd = autoStep()
        autoSub = autoStep(false)

        addView()
    }

    private fun autoStep(add: Boolean = true) : Runnable {
        var runnable: Runnable? = null
        runnable = Runnable {
            autoExecuted = true
            onAction(add)
            postDelayed(runnable, if (++autoCount > 30) 25 else 55.toLong())
        }
        return runnable
    }


    fun setStep(step: BigDecimal): NumberModifierView {
        this.step = step
        return this
    }

    fun allowNegatives(): NumberModifierView {
        allowNegatives = true
        return this
    }

    private fun addView() {
        val subCv = createCardView()
        val params = subCv.layoutParams as LayoutParams
        params.marginEnd = FormatUtils.toDp(resources.displayMetrics, 3)
        subCv.addView(createImageView(R.drawable.ic_substract_24dp))
        modifyEditText(subCv, false)
        addView(subCv)

        val addCv = createCardView()
        addCv.addView(createImageView(R.drawable.ic_add_24dp))
        modifyEditText(addCv, true)
        addView(addCv)
    }

    private fun createCardView(): CardView {
        val cardView = CardView(context)
        val displayMetrics = resources.displayMetrics

        val size = FormatUtils.toDp(displayMetrics, 50)
        val layoutParams = LayoutParams(size, size)
        cardView.layoutParams = layoutParams

        val color = context.resources.getColor(
            R.color.dark,
            context.theme)

        cardView.setCardBackgroundColor(color)
        cardView.radius = FormatUtils.toDpFloat(displayMetrics, 3)
        cardView.cardElevation = 0f

        return cardView
    }

    private fun createImageView(@DrawableRes drawable: Int): ImageView {
        val imageView = ImageView(context)

        val size = FormatUtils.toDp(resources.displayMetrics, 22)
        val layoutParams = FrameLayout.LayoutParams(size, size)
        layoutParams.gravity = Gravity.CENTER
        imageView.layoutParams = layoutParams

        imageView.setImageResource(drawable)

        val color = context.resources.getColor(
            R.color.white,
            context.theme)
        imageView.setColorFilter(color)

        return imageView
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun modifyEditText(cardView: CardView, addition: Boolean) {
        cardView.setOnTouchListener { _, event: MotionEvent ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    if (addition) {
                        removeCallbacks(autoAdd)
                        autoCount = 0
                        postDelayed(autoAdd, 300)
                    } else {
                        removeCallbacks(autoSub)
                        autoCount = 0
                        postDelayed(autoSub, 300)
                    }
                    return@setOnTouchListener true
                }
                MotionEvent.ACTION_UP -> {
                    removeCallbacks(if (addition) autoAdd else autoSub)
                    if (autoExecuted) {
                        autoExecuted = false
                    } else {
                        onAction(addition)
                    }
                    return@setOnTouchListener true
                }
            }
            false
        }
    }

    private fun onAction(addition: Boolean) {
        val editText: EditText? = rootView.findViewById(targetId)
        if (editText != null) {
            var value = editText.bigDecimal
            val step = if (addition) step else step.negate()

            value = value.add(step)
            if (!allowNegatives && value <= BigDecimal.ZERO)
                value = BigDecimal.ZERO

            editText.bigDecimal = value
            editText.setSelection(editText.text.length)
        }
    }
}